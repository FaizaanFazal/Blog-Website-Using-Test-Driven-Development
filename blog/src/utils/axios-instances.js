// import axios from 'axios';

// export const axiosInstance = axios.create({
//   baseURL: 'http://localhost:8000/',
// });
import axios from 'axios';

const placeholderApi = axios.create({
  baseURL: 'http://localhost:8000',
});

export default placeholderApi;
