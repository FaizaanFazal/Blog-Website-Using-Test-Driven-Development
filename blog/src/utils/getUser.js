export const getUser = () => {
    const userString = localStorage.getItem('user');
    if (userString) {
      const user = JSON.parse(userString);
      return user;
    } else {
      return null;
    }
  };