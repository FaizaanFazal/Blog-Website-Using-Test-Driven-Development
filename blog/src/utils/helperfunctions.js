import { fireEvent } from '@testing-library/react';
import images from './images';

export const isProperImageURL = (url) => {
  /* eslint-disable-next-line no-useless-escape */
  const regex = /^(https?|ftp):\/\/(([a-z\d]([a-z\d-]*[a-z\d])?\.)+[a-z]{2,}|localhost)(\/[-a-z\d%_.~+]*)*(\?[;&a-z\d%_.~+=-]*)?(\#[-a-z\d_]*)?$/i;
  return regex.test(url);
};
export const isProperEmail = (email) => {
  const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return emailRegex.test(email);
};
export const isProperPass = (pass) => {
  const passregex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$/;
  return passregex.test(pass);
};
export function backspace(element) {
  let actuallyTyped = element.value;

  const backspaceKey = {
    key: 'Backspace',
    code: 8,
    inputType: 'deleteContentBackward',
  };

  const sharedEventConfig = {
    key: backspaceKey.key,
    charCode: backspaceKey.code,
    keyCode: backspaceKey.code,
    which: backspaceKey.code,
    modifier: backspaceKey.modifier,
  };
  const downEvent = fireEvent.keyDown(element, sharedEventConfig);

  if (downEvent) {
    actuallyTyped = actuallyTyped.slice(0, -1);

    fireEvent.input(element, {
      target: { value: actuallyTyped },
      inputType: backspaceKey.inputType,
      bubbles: true,
      cancelable: true,
    });
  }

  fireEvent.keyUp(element, sharedEventConfig);
}

export const small = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: 'Elena Rodriguez' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    verified: true,
    slug: 'post1',
    content: "test Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.",
  },
];
export const greater = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: ' Elena' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    slug: 'post1',
    verified: true,
    content: "Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n",
  }];
