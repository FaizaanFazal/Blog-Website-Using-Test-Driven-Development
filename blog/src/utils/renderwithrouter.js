import {render} from '@testing-library/react'
import * as userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom'
import {MemoryRouter} from 'react-router-dom'

export const renderWithRouter = (ui, {route = '/'} = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    user: userEvent.setup,
    ...render(ui, {wrapper: MemoryRouter}),
  }
}