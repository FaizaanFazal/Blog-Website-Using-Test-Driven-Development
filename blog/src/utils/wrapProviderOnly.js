import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

const AllTheProviders = ({ children }) => (
  <Provider store={reduxStore}>
    {children}
  </Provider>
);

export const renderWithProvidersOnly = (ui, options) => render(<AllTheProviders>{ui}</AllTheProviders>, options);

// export function renderWithProviders(
//   ui,
//   {
//     // preloadedState = {},
//     // Automatically create a store instance if no store was passed in
//     store = reduxStore,
//     ...renderOptions
//   } = {},
// ) {
//   function Wrapper({ children }) {
//     return <Provider store={store}>{children}</Provider>;
//   }

//   // Return an object with the store and all of RTL's query functions
//   return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
// }

renderWithProvidersOnly.propTypes = {
  children: PropTypes.node.isRequired,
};
