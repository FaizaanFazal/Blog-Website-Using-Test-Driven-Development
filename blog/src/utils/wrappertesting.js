import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

export function renderWithProviders(
  ui,
  {
    // preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = reduxStore,
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

renderWithProviders.propTypes = {
  children: PropTypes.node.isRequired,
};
