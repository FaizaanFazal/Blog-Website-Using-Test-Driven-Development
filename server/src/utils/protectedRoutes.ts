import { authenticateToken } from "@/models/authModel";
import { Request, Response, NextFunction, RequestHandler } from "express";


export const protectedRoutes = [
    {
        method: "",
        path: ""
    },
    {
        method: "",
        path: ""
    }];


//check protected routes to check authentication
export const isProtectedRouteMiddleware: RequestHandler = (req: Request, res: Response, next: NextFunction) => {
    console.log(req.method)
    const protectedPathData = protectedRoutes.find((route) => route.path === req.path);
    if (protectedPathData) {
        const authenticatedUser = authenticateToken(req, res, next);
        if (authenticatedUser !== 403 && authenticatedUser !== 401) {
            req.user = authenticatedUser;
            console.log("User", authenticatedUser)
        } else {
            res.status(authenticatedUser);
            res.end();
        }
    } else {
        next();
    }
};