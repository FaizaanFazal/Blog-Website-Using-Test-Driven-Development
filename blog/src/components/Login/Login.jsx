import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { ToastContainer, toast } from 'react-toastify';
import { isProperEmail, isProperPass } from '../../utils/helperfunctions';

export default function Login() {
  const [email, setEmail] = useState('');
  const [pass, setPass] = useState('');
  const [errorEmail, setErrorEmail] = useState('');
  const [errorPass, setErrorPass] = useState('');

  const emailValidation = (e) => {
    setEmail(e.target.value);
    const { value } = e.target;
    const check = isProperEmail(value);
    if (check === false) { setErrorEmail('* Email Format incorrect.'); } else { setErrorEmail(null); }
  };

  const passValidation = (e) => {
    setPass(e.target.value);
    const { value } = e.target;
    const check = isProperPass(value);
    if (check === false) { setErrorPass('* Required min 8 characters.*One Special char,*On Capital letter.*One digit'); } else { setErrorPass(null); }
  };

  const validateAll = () => {
    if (email.length < 8 || !isProperEmail(email)) {
      toast.error('Email incorrect', { position: toast.POSITION.TOP_RIGHT, duration: 5000 });
      return false;
    }
    if (!isProperPass(pass)) {
      toast.error('Password Incorrect', { position: toast.POSITION.TOP_RIGHT, duration: 5000 });
      return false;
    }
    return true;
  };

  const handlelogin = (e) => {
    e.preventDefault();
    const checkall = validateAll();
    if (checkall === false) {
      return false;
    }
    alert('validations working fine');
    return true;
  };

  const areFieldfilled = () => {
    if (!email || !pass) {
      return false;
    }
    return true;
  };

  return (
    <section data-testid="loginForm">
      <ToastContainer />
      <div className="container">
        <div className="box-auth bg-white">
          <div className="heading">
            <h4 className="title title-box text-center">Login</h4>
          </div>

          <form className="text-darkblue p-3">
            <div className="form-group">
              <label>Email</label>
              <input
                type="email"
                data-testid="inputEmail"
                className="form-control"
                placeholder="Title..."
                onChange={emailValidation}
              />
              {errorEmail && <small data-testid="errorEmail" style={{ color: 'red' }}>{errorEmail}</small>}
            </div>

            <div className="form-group">
              <label>Password</label>
              <input
                type="password"
                data-testid="inputPass"
                className="form-control"
                placeholder="Password..."
                onChange={passValidation}
              />
              {errorPass && <small data-testid="errorPass" style={{ color: 'red' }}>{errorPass}</small>}
            </div>

            <div className="text-center">
              <button
                data-testid="loginbtn"
                type="submit"
                className="btn btn-primary rounded-pill"
                onClick={handlelogin}
                disabled={!areFieldfilled()}
              >
                Login
              </button>
            </div>
            <div className="d-flex justify-content-between">
              <small><Link to="/forgetpass">Forget password?</Link></small>
              <small><Link to="/signup">create an account!</Link></small>
            </div>

          </form>
        </div>

      </div>
    </section>
  );
}
