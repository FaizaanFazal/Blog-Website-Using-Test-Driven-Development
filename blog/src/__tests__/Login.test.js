import React from 'react';
import { fireEvent, screen,userEvent } from '@testing-library/react';
import { renderWithProviders } from '../utils/wrappertesting';
import Login from '../components/Login/Login';
import { backspace } from '../utils/helperfunctions';
import { debug } from 'console';

describe('Unit tests for Login Components', () => {
  it('Snapshot test for Login component', () => {
    const { asFragment } = renderWithProviders(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Dom test for-> renders the Login', () => {
    renderWithProviders(<Login />);
    const loginForm = screen.getByTestId('loginForm');
    expect(loginForm).toBeInTheDocument();
  });

  it('Email input changes on input', () => {
    renderWithProviders(<Login />);
    const emailid = screen.getByTestId('inputEmail');
    fireEvent.change(emailid, { target: { value: 'test@gmail.com' } });
    expect(emailid.value).toBe('test@gmail.com');
  });

  it('Password input changes on input and has type password', () => {
    renderWithProviders(<Login />);
    const pass = screen.getByTestId('inputPass');
    fireEvent.change(pass, { target: { value: 'qwe123' } });
    expect(pass.value).toBe('qwe123');
    expect(pass.type).toBe('password');
  });

  it('Login Button', () => {
    renderWithProviders(<Login />);
    const loginbtn = screen.getByTestId('loginbtn');
    expect(loginbtn).toBeInTheDocument();
  });

  it('Validations Checking on Email', () => {
    renderWithProviders(<Login />);
    const inputEmail = screen.getByTestId('inputEmail');
    fireEvent.change(inputEmail, { target: { value: 'faizaan@gmailcom' } }); // wrong email format should raise error
    const errorEmail = screen.getByTestId('errorEmail');
    expect(errorEmail).toBeInTheDocument();
    fireEvent.change(inputEmail, { target: { value: 'faizaan@gmail.com' } }); // error should not be there
    expect(errorEmail).not.toBeInTheDocument();
    fireEvent.keyDown(inputEmail, 'Backspace');
    fireEvent.keyUp(inputEmail, 'Backspace'); // removes last character
    fireEvent.keyDown(inputEmail, 'Backspace');
    fireEvent.keyUp(inputEmail, 'Backspace'); // removes last character
    fireEvent.keyDown(inputEmail, 'Backspace');
    fireEvent.keyUp(inputEmail, 'Backspace'); // removes last character
    screen.debug()
    expect(errorEmail).toBeInTheDocument();
    
  });

  it('Validations Checking on Password', () => {
    renderWithProviders(<Login />);
    const inputPass = screen.getByTestId('inputPass');
    fireEvent.change(inputPass, { target: { value: 'qwe123' } }); // less than 8 character, no special/capital letter error shoul be in document
    const errorPass = screen.getByTestId('errorPass');
    expect(errorPass).toBeInTheDocument();
    fireEvent.change(inputPass, { target: { value: 'Qwer@124' } }); // conditions satified should not be there
    expect(errorPass).not.toBeInTheDocument();

    backspace(inputPass);
    userEvent.keyboard('{Escape}');
    // remove last character  min chars become 7
    expect(errorPass).toBeInTheDocument();
  });
});
