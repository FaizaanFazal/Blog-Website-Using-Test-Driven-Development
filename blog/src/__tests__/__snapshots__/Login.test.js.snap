// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for Login Components Snapshot test for Login component 1`] = `
<DocumentFragment>
  <section
    data-testid="loginForm"
  >
    <div
      class="Toastify"
    />
    <div
      class="container"
    >
      <div
        class="box-auth bg-white"
      >
        <div
          class="heading"
        >
          <h4
            class="title title-box text-center"
          >
            Login
          </h4>
        </div>
        <form
          class="text-darkblue p-3"
        >
          <div
            class="form-group"
          >
            <label>
              Email
            </label>
            <input
              class="form-control"
              data-testid="inputEmail"
              placeholder="Title..."
              type="email"
            />
          </div>
          <div
            class="form-group"
          >
            <label>
              Password
            </label>
            <input
              class="form-control"
              data-testid="inputPass"
              placeholder="Password..."
              type="password"
            />
          </div>
          <div
            class="text-center"
          >
            <button
              class="btn btn-primary rounded-pill"
              data-testid="loginbtn"
              disabled=""
              type="submit"
            >
              Login
            </button>
          </div>
          <div
            class="d-flex justify-content-between"
          >
            <small>
              <a
                href="/forgetpass"
              >
                Forget password?
              </a>
            </small>
            <small>
              <a
                href="/signup"
              >
                create an account!
              </a>
            </small>
          </div>
        </form>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
